### PR TITLE
Switch Jaeger remote sampler to use grpc lite

### LINKED ISSUE #4014: Switch Jaeger remote sampler to "grpc" lite mechanism
**Is your feature request related to a problem? Please describe.**

This is required for https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/4898#issuecomment-995178433

I am starting to work on this

### PR DESCRIPTION
**Pull request by pavolloffay**
Resolves #4014 

Notable changes:

- added some methods to `CodedInpotStream` to support additional data types
- Jaeger remote sampler support only grpc light via okhttp
- switched tests from in-process grpc to armeria

### CODE REVIEW

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/OkHttpGrpcService.java
Comment (anuraaga) [inline sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/OkHttpGrpcService.java:38]: Let's keep a static reference to the real logger

File: sdk-extensions/jaeger-remote-sampler/build.gradle.kts
Comment (anuraaga) [inline sdk-extensions/jaeger-remote-sampler/build.gradle.kts:9]: I don't think we need as many as the OTLP tests but can we add one test suite with the grpc dependencies used?

https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/otlp/trace/build.gradle.kts#L31

It's ok to just copy the test code instead of abstract

**Review by anuraaga — COMMENTED** [review]
Thanks a lot for this, looks great! Hope you could learn something about gRPC and enjoy it while writing the PR. If not, then good job anyways :)

File: sdk-extensions/jaeger-remote-sampler/build.gradle.kts
Comment (pavolloffay) [reply sdk-extensions/jaeger-remote-sampler/build.gradle.kts:9]: I have added support only for "grpc light" approach via okhttp.

**Comment by pavolloffay** [comment]
@anuraaga the coverage is low, but it seems like the "grpc lite" classes in `otlp:common` are not covered either.

**Comment by anuraaga** [comment]
@pavolloffay Since you added the MarshallerRemoteSamplerServiceGrpc I noticed it's pretty easy to just go ahead and support okhttp and grpc like our exporters so I added https://github.com/open-telemetry/opentelemetry-java/pull/4043/commits/a31e6c05b1261a9d36b9b062c565d55fe2fc6eab to do so, please check it out. We can deprecate setChannel later across all the packages.

File: exporters/otlp/common/src/main/java/io/opentelemetry/exporter/otlp/internal/grpc/OkHttpGrpcExporter.java
Comment (anuraaga) [inline exporters/otlp/common/src/main/java/io/opentelemetry/exporter/otlp/internal/grpc/OkHttpGrpcExporter.java:216]: Is this change needed?

**Review by anuraaga — COMMENTED** [review]
> @anuraaga the coverage is low, but it seems like the "grpc lite" classes in otlp:common are not covered either.

otlp:common all folders have 80%+ coverage and this one has 66% now so not sure they're comparable :P

Can you add some tests for the error cases, etc and get the coverage up? This artifact hasn't really had an owner so has inherited low test coverage such as no test of certain samplers, any sort of server error condition. Since you're trying to roll out jaeger remote sampler across OTel I presume it's ok to think of you as an owner :) If you can help add test coverage would be great, if you want to do it in another PR it's fine but would need to actually do that please.

File: exporters/otlp/common/src/main/java/io/opentelemetry/exporter/otlp/internal/grpc/OkHttpGrpcExporter.java
Comment (pavolloffay) [reply exporters/otlp/common/src/main/java/io/opentelemetry/exporter/otlp/internal/grpc/OkHttpGrpcExporter.java:216]: yes, otherwise I would have to copy the whole method.

Comment (pavolloffay) [reply exporters/otlp/common/src/main/java/io/opentelemetry/exporter/otlp/internal/grpc/OkHttpGrpcExporter.java:216]: copying it for now

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/DefaultGrpcServiceBuilder.java
Comment (jack-berg) [inline sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/DefaultGrpcServiceBuilder.java]: Let's extract this to a helper function now that it appears in a couple of places.

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/OkHttpGrpcService.java
Comment (jack-berg) [inline sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/OkHttpGrpcService.java:49]: This is unused.

File: exporters/otlp/common/src/main/java/io/opentelemetry/exporter/otlp/internal/CodedInputStream.java
Comment (pavolloffay) [reply exporters/otlp/common/src/main/java/io/opentelemetry/exporter/otlp/internal/CodedInputStream.java:206]: this can be done in a separate PR

Comment (anuraaga) [reply exporters/otlp/common/src/main/java/io/opentelemetry/exporter/otlp/internal/CodedInputStream.java:206]: Let's do this simple typo correction in this PR

Comment (pavolloffay) [reply exporters/otlp/common/src/main/java/io/opentelemetry/exporter/otlp/internal/CodedInputStream.java]: done

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/UnMarshaller.java
Comment (pavolloffay) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/UnMarshaller.java]: done

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/DefaultGrpcServiceBuilder.java
Comment (pavolloffay) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/DefaultGrpcServiceBuilder.java]: done

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/GrpcService.java
Comment (pavolloffay) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/GrpcService.java]: +1 on changing return type to void. 

changed also to `OkHttpGrpcService<ReqMarshalerT extends Marshaler, ResUnMarshalerT extends UnMarshaler>`  which looks weirder than before IMHO

**Review by anuraaga — APPROVED** [review]
Thanks @pavolloffay - I have added back code from my commit. I had to revert the change of the method signature to `void`. There is some more cleanup that can be done but that can happen later

**Comment by pavolloffay** [comment]
Do not merge yet, the tests are flaky.

**Comment by pavolloffay** [comment]
I am going to look at fixing them. 

regarding the return type of the service, it would be appropriate propagate the result of the call to the caller.

**Comment by pavolloffay** [comment]
@anuraaga I could not reproduce the failing test (`unimplemented_error_server_response`) locally. 

```
while ./gradlew --no-build-cache  :sdk-extensions:jaeger-remote-sampler:cleanTest :sdk-extensions:jaeger-remote-sampler:check      ; do :; done  
```

Now it seems that it is passing on CI.

**Comment by pavolloffay** [comment]
The second to last CI job failed on the flaky test. @anuraaga any hints why the logs are not being captured? 

```
JaegerRemoteSamplerTest > unimplemented_error_server_response() FAILED
    org.opentest4j.AssertionFailedError: Contain the string <Server responded with UNIMPLEMENTED.> ==> None of the 7 captured log events matched the filter predicate
        at app//org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:39)
        at app//org.junit.jupiter.api.Assertions.fail(Assertions.java:134)
        at app//io.github.netmikey.logunit.api.LogCapturer.lambda$assertContains$3(LogCapturer.java:174)
        at java.base@17.0.1/java.util.Optional.orElseGet(Optional.java:364)
        at app//io.github.netmikey.logunit.api.LogCapturer.assertContains(LogCapturer.java:173)
        at app//io.github.netmikey.logunit.api.LogCapturer.assertContains(LogCapturer.java:121)
        at app//io.opentelemetry.sdk.extension.trace.jaeger.sampler.JaegerRemoteSamplerTest.unimplemented_error_server_response(JaegerRemoteSamplerTest.java:351)
```

**Comment by anuraaga** [comment]
@pavolloffay I added another commit https://github.com/open-telemetry/opentelemetry-java/pull/4043/commits/dfff2d71226348cede4d78e16989878be281c20c to workaround issue with logunit. We've generally been able to avoid problems by using `private static final Logger` in classes (previous issues were when we had `private final Logger`). Not sure what's going on this time but worked around it anyways

**Comment by anuraaga** [comment]
Hmm had confidence in my workaround but somehow it's not working, need to look more

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/SamplingStrategyResponse.java
Comment (jack-berg) [inline sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/SamplingStrategyResponse.java:30]: All the methods can can have their visibility reduced to package private

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/UnMarshaler.java
Comment (jack-berg) [inline sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/UnMarshaler.java:18]: No need for a public method on a package private class.

**Review by jack-berg — COMMENTED** [review]
Couple of additional comments. Looking pretty good though 👍

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/DefaultGrpcServiceBuilder.java
Comment (anuraaga) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/DefaultGrpcServiceBuilder.java:124]: I think in a soon followup we can add all except `addRetryPolicy` to the builder. I don't think it hurts too much keeping them.

`setChannel` we hope is not used but since it was public let's stick with having it for now and deprecating all of our `setChannel` methods in one PR

**Comment by anuraaga** [comment]
@pavolloffay FYI I added some commits, please don't lose them :)

For the record, I will be on the hook if the build is flaky after merging this. 

I am hopeful for https://github.com/open-telemetry/opentelemetry-java/pull/4043/commits/d5a38b5d105c1a311c4d20305fd0d3c68d4f5be4 my rough hypothesis is that due to HTTP/2 details, it depends on the state of the connection if there is an empty body sent and the trailer includes the status, or the headers include the status and there's no body and body read fails, which previously skipped our status handling logic.

**Review by jack-berg — APPROVED** [review]

**Comment by anuraaga** [comment]
Thanks @pavolloffay for helping with this tedious PR!

**Comment by pavolloffay** [comment]
it was fun to look at the proto and grcp! Thanks for guidance as well :P

**Comment by 4orty** [comment]
@anuraaga 
I'm encountered "could not consume server response" error. Can you explain about this for us?
I'm using opentelemetry java SDK 1.49.0 and okhttp 4.12.0

Is it related with state of the connection?

**Comment by jack-berg** [comment]
Please open a new issue to discuss/ track

**Pull request merged**
